00_Base/src/datasources/TokensDatasource.ts

```diff
@@ -0,0 +1,316 @@
+        mappedOcppAuth,
+        {
+          idToken: tokenDto.uid,
+          type: OcpiTokensMapper.mapOcpiTokenTypeToOcppIdTokenType(
+            tokenDto.type,
+          ),
+        },
+      );
+
+    if (!savedOcppAuth) {
+      throw new UnknownTokenException(
+        'Authorization could not be created upon saving token',
+      );
+    }
+
+    try {
+      const ocpiToken = await this.createOrUpdateOcpiToken(
+        savedOcppAuth?.id,
+        tokenDto,
+      );
+      return OcpiTokensMapper.toDto(savedOcppAuth!, ocpiToken);
+    } catch (error) {
+      this.logger.error('Error saving token', error);
+      throw error;
+    }
+  }
+
+  async getOcpiTokenByAuthorizationId(
+    authorizationId: string,
+  ): Promise<OcpiToken | undefined> {
+    const ocpiToken = await this.tokensRepository.readOnlyOneByQuery({
+      where: {
+        authorization_id: authorizationId,
+      },
+    });
+    if (!ocpiToken) {
+      // TODO better error
+      return Promise.resolve(undefined);
+    }
+    return Promise.resolve(ocpiToken);
+  }
+
+  private getMatchingAuth(
+    id: number,
+    ocppAuths: Authorization[],
+  ): Authorization | undefined {
+    return ocppAuths.find((ocppAuth) => ocppAuth.id === id);
+  }
+
+  private async createOrUpdateOcppAuth(
+    tokenDto: TokenDTO,
+    transaction?: SequelizeTransaction,
+  ): Promise<Authorization> {
+    const ocppAuth = OcpiTokensMapper.mapOcpiTokenToOcppAuthorization(tokenDto);
+    const queryCondition = {
+      idToken: tokenDto.uid,
+      type: OcpiTokensMapper.mapOcpiTokenTypeToOcppIdTokenType(tokenDto.type),
+    };
+
+    const savedAuth =
+      await this.authorizationRepository.createOrUpdateByQuerystring(
+        ocppAuth,
+        queryCondition,
+        transaction,
+      );
+
+    if (!savedAuth) {
+      throw new UnknownTokenException(
+        'Authorization could not be created upon saving token',
+      );
+    }
+
+    return await savedAuth.reload({
+      include: [IdToken, IdTokenInfo],
+      transaction,
+    });
+  }
+
+  private async createOrUpdateOcpiToken(
+    authorizationId: number,
+    tokenDto: TokenDTO,
+  ): Promise<OcpiToken> {
+    const baseTokenData = {
+      visual_number: tokenDto.visual_number,
+      issuer: tokenDto.issuer,
+      whitelist: tokenDto.whitelist,
+      default_profile_type: tokenDto.default_profile_type,
+      energy_contract: tokenDto.energy_contract,
+      last_updated: tokenDto.last_updated,
+      type: tokenDto.type,
+    };
+
+    const [savedOcpiToken, ocpiTokenCreated] =
+      await this.tokensRepository.readOrCreateByQuery({
+        where: {
+          authorization_id: authorizationId,
+        },
+        defaults: {
+          ...baseTokenData,
+          authorization_id: authorizationId,
+          country_code: tokenDto.country_code,
+          party_id: tokenDto.party_id,
+        },
+      });
+
+    // TODO confirm all updatable properties
```

Comment:
@ChrisWeissmann correct me if wrong but I think this TODO can be removed as all props are updatable.

Reply:
Works for me.